File: arrays/20_valid_parentheses.py
"""
20. Valid Parentheses
Easy

8216

337

Add to List

Share
Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
 

Example 1:

Input: s = "()"
Output: true
Example 2:

Input: s = "()[]{}"
Output: true
Example 3:

Input: s = "(]"
Output: false
Example 4:

Input: s = "([)]"
Output: false
Example 5:

Input: s = "{[]}"
Output: true
 

Constraints:

1 <= s.length <= 104
s consists of parentheses only '()[]{}'.
Accepted
1,526,239
Submissions
3,792,935
"""

"""
Runtime: 28 ms, faster than 86.43% of Python3 online submissions for Valid Parentheses.
Memory Usage: 14.4 MB, less than 34.15% of Python3 online submissions for Valid Parentheses.
"""

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        opening_brackets = {'(', '{', '['}
        for char in s:
            if char in opening_brackets:
                stack.append(char)
            elif char == ')' and stack and stack[-1] == '(':
                stack.pop()
            elif char == '}' and stack and stack[-1] == '{':
                stack.pop()
            elif char == ']' and stack and stack[-1] == '[':
                stack.pop()
            else:
                return False
        
        return not stack

File: algo/20_valid_parentheses.go
package main

/*
Runtime: 0ms Beats 100% of users with GoLang
Memory Usage: 4.42 MB Beats 15.76% of users with GoLang
*/
func isValid(s string) bool {
	characters := []rune(s)

	stack := []rune{}

	for _, char := range characters {

		switch char {

		case ')':
			if len(stack) != 0 && stack[len(stack)-1] == '(' {
				stack = stack[:len(stack)-1]
				continue
			}

		case '}':
			if len(stack) != 0 && stack[len(stack)-1] == '{' {
				stack = stack[:len(stack)-1]
				continue
			}

		case ']':
			if len(stack) != 0 && stack[len(stack)-1] == '[' {
				stack = stack[:len(stack)-1]
				continue
			}
		}

		stack = append(stack, char)

	}

	return len(stack) == 0

}
